### PR TITLE
[cloud-provider-azure] go refactoring

### PR DESCRIPTION
**Pull request by vitaliy-sn**
## Description

This MR is a part of "Global" Deckhouse refactoring.

## Why do we need it, and what problem does it solve?

It is aimed to improve UX and DX.

Go is more reliable and suitable to describe cloud infrastructure components, especially in distributed systems/environments
Open API specs will help us with bounding a contract between user and deckhouse, and between hooks and helm
An Open API spec for a module will become the single source of truth (documenting, validating, testing)
More tests, more stability

## Changelog entries

```changes
section: cloud-provider-azure
type: chore
summary: Rewrite hooks on golang
```

### CODE REVIEW

File: modules/030-cloud-provider-azure/hooks/azure_cluster_configuration.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+		return nil, fmt.Errorf("cannot convert kubernetes object: %v", err)
+	}
+
+	return secret, nil
+}
+
+func clusterConfiguration(input *go_hook.HookInput) error {
+	secret := input.Snapshots["provider_cluster_configuration"][0].(*v1.Secret)
+
+	clusterConfigurationYAML := secret.Data["cloud-provider-cluster-configuration.yaml"]
```
Comment (name212) [inline modules/030-cloud-provider-azure/hooks/azure_cluster_configuration.go:93]: you can use helper hook https://github.com/deckhouse/deckhouse/blob/main/go_lib/hooks/cluster_configuration/hook.go

Comment (name212) [reply modules/030-cloud-provider-azure/hooks/azure_cluster_configuration.go:93]: i see that bash hook is simple. It copy provider cluster configuration and discovery cloud data.
And i do not see usage structures in another hooks.
I think we simplify this hook like in [yandex](https://github.com/deckhouse/deckhouse/blob/main/modules/030-cloud-provider-yandex/hooks/yandex_cluster_configuration.go)

File: modules/030-cloud-provider-azure/hooks/azure_cluster_configuration_test.go [hunk 2]
```diff
@@ -159,44 +134,22 @@ data:
...
+			Expect(f.ValuesGet("cloudProviderAzure.internal.providerDiscoveryData.instances.diskType").String()).To(Equal("test"))
+			Expect(f.ValuesGet("cloudProviderAzure.internal.providerDiscoveryData.instances.additionalTags").String()).To(Equal("{}"))
 
-			Expect(f.ValuesGet("cloudProviderAzure.internal.providerClusterConfiguration.masterNodeGroup.instanceClass.additionalTags.node").String()).To(Equal("master"))
-
-			Expect(f.ValuesGet("cloudProviderAzure.internal.providerClusterConfiguration.nodeGroups.0.name").String()).To(Equal("static"))
-			Expect(f.ValuesGet("cloudProviderAzure.internal.providerClusterConfiguration.nodeGroups.0.replicas").String()).To(Equal("1"))
-			Expect(f.ValuesGet("cloudProviderAzure.internal.providerClusterConfiguration.nodeGroups.0.zones").String()).To(MatchJSON(`["1","2","3"]`))
-
-			Expect(f.ValuesGet("cloudProviderAzure.internal.providerClusterConfiguration.nodeGroups.0.instanceClass.machineSize").String()).To(Equal("Standard_F2"))
```
Comment (name212) [inline modules/030-cloud-provider-azure/hooks/azure_cluster_configuration_test.go:185]: why we remove this tests?

File: modules/030-cloud-provider-azure/hooks/storage_classes.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		Name: "managed-premium",
+		Type: "Premium_LRS",
+	},
+}
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 20},
+}, storageClasses)
+
+func storageClasses(input *go_hook.HookInput) error {
```
Comment (name212) [inline modules/030-cloud-provider-azure/hooks/storage_classes.go:53]: Can we use library [hook](https://github.com/deckhouse/deckhouse/blob/main/go_lib/hooks/storage_class/hook.go) here?

File: modules/030-cloud-provider-azure/hooks/wait_for_all_master_nodes_to_become_ready.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
+/*
```
Comment (name212) [inline modules/030-cloud-provider-azure/hooks/wait_for_all_master_nodes_to_become_ready.go:1]: we have library [hook](https://github.com/deckhouse/deckhouse/blob/main/go_lib/hooks/nodes/hook.go#L34) for it

Can we use it here?

File: modules/030-cloud-provider-azure/openapi/values.yaml [hunk 5]
```diff
@@ -0,0 +1,79 @@
+x-extend:
+  schema: config-values.yaml
+type: object
```
Comment (name212) [inline modules/030-cloud-provider-azure/openapi/values.yaml:3]: I think better way use links to cloud-provides specs for exclude copy-paste like [yandex](https://github.com/deckhouse/deckhouse/blob/main/modules/030-cloud-provider-yandex/openapi/values.yaml#L19).

Comment (name212) [reply modules/030-cloud-provider-azure/openapi/values.yaml:3]: https://github.com/deckhouse/deckhouse/blob/main/candi/cloud-providers/yandex/openapi/cluster_configuration.yaml#L7

File: modules/030-cloud-provider-azure/hooks/azure_cluster_configuration_test.go (hunk 2 above)
Comment (vitaliy-sn) [reply modules/030-cloud-provider-azure/hooks/azure_cluster_configuration_test.go:185]: These values are only used in terraform, so there is no point in testing them here.

File: modules/030-cloud-provider-azure/hooks/storage_classes.go (hunk 3 above)
Comment (vitaliy-sn) [reply modules/030-cloud-provider-azure/hooks/storage_classes.go:53]: We cannot use it without modifications. This library is currently used only in yandex provider.

**Review by z9r5 — DISMISSED** [review]

File: modules/030-cloud-provider-azure/hooks/azure_cluster_configuration_test.go (hunk 2 above)
Comment (name212) [reply modules/030-cloud-provider-azure/hooks/azure_cluster_configuration_test.go:185]: ok

File: modules/030-cloud-provider-azure/hooks/storage_classes.go (hunk 3 above)
Comment (name212) [reply modules/030-cloud-provider-azure/hooks/storage_classes.go:69]: ok.

**Review by name212 — CHANGES_REQUESTED** [review]
Lint error and DCO

**Review by name212 — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @konstantin-axenov. I've started the workflow for label 'e2e/run/azure'...

**Review by konstantin-axenov — DISMISSED** [review]

File: modules/030-cloud-provider-azure/hooks/wait_for_all_master_nodes_to_become_ready.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+Copyright 2021 Flant JSC
```
Comment (konstantin-axenov) [inline modules/030-cloud-provider-azure/hooks/wait_for_all_master_nodes_to_become_ready.go:2]:
```suggestion
Copyright 2022 Flant JSC
```

File: modules/030-cloud-provider-azure/hooks/azure_cluster_configuration.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2021 Flant JSC
```
Comment (konstantin-axenov) [inline modules/030-cloud-provider-azure/hooks/azure_cluster_configuration.go:2]:
```suggestion
Copyright 2022 Flant JSC
```

File: modules/030-cloud-provider-azure/hooks/storage_classes.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2021 Flant JSC
```
Comment (konstantin-axenov) [inline modules/030-cloud-provider-azure/hooks/storage_classes.go:2]:
```suggestion
Copyright 2022 Flant JSC
```

**Review by konstantin-axenov — APPROVED** [review]

**Pull request merged**